### If you want to play with some genetics code chunks

In your R Console (the bottom left panel), enter the following commands to 
install a plotting package that we will now use:
  
  ```
install.packages("ggplot2")
```

The code below simulates genotyping and gene expression data, which we can then 
test for an eQTL

```{r simulated-data}

## Simulate 10 SNPs with different minor allele frequencies for 300 individuals
set.seed(100)
n <- 300
maf <- c(0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5)
prob <- cbind((1-maf)^2, 2*maf*(1-maf), maf^2)
geno <- apply(prob, 1, function(p) sample(0:2, size=n, prob=p, replace=TRUE))
geno <- data.frame(paste("sample", 1:n, sep="_"), geno)
names(geno) <- c("sample", paste("snp", 1:(ncol(geno)-1), sep="_"))

## Simulate gene expression for one gene per SNP.
## Expression is base + snp + noise
set.seed(101)
snp <- geno
errorMeans <- rnorm(nrow(snp))
expr <- sapply(snp[-1], function(g) 7 + 1.5*g) + 
  matrix(rnorm(nrow(snp)*(ncol(snp)-1), mean=errorMeans, sd=0.5), nrow=nrow(snp))
expr <- data.frame(snp$sample, expr)
names(expr) <- c("sample", paste("gene", 1:(ncol(expr)-1), sep="_"))
```

### Plotting gene expression by genotyping

A convenient way to display gene expression values by genotype is as box plots. These provide a good, non-parametric, indication of the distributions. To convey a sense of the frequency of each genotype in the sample it is useful to also add points for each individual to the plot. Below is an example of how this might look for each of the ten SNP/gene pairs.

```{r boxplots}
library(ggplot2)
genoLong <- tidyr::gather(geno, snp, genotype, -sample)
exprLong <- tidyr::gather(expr, gene, expression, -sample)
dataLong <- cbind(genoLong, exprLong["expression"])
dataLong$genotype <- as.factor(dataLong$genotype) 
ggplot(dataLong, aes(genotype, expression)) +
  geom_jitter(colour="darkgrey", position=position_jitter(width=0.25)) +
  geom_boxplot(outlier.size=0, alpha=0.6, fill="grey") + 
  facet_wrap(~snp) + theme_bw()
```

### Estimating SNP effects

To obtain estimates of the genotypic contribution to gene expression we fit a simple linear regression model of the form $E_i = \beta_0 + \beta G_i + \varepsilon$, where $E_i$ is the vector of gene expression values for gene $i$ and $G_i$ is the genotype vector for SNP $i$. We are interested in the estimate for $\beta$ which indicates the change in gene expression for each copy of the second allele.

```{r eqtl}
fit <- mapply(function(e, g) lm(e ~ g), expr[-1], geno[-1], SIMPLIFY=FALSE)
betaHat <- sapply(fit, coef)[2,]
betaHat 
```

We use the function confint to obtain 95% confidence intervals of the estimated SNP effects.

```{r CI}
ci <- sapply(fit, confint, "g")
rownames(ci) <- c("lower", "upper")
ci
```